File: Python/450. Delete Node in a BST.py
'''
1.递归
2.迭代
'''


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def deleteOneNode(self, node):
#         if not node.left and not node.right:
#             return None
#         elif not node.left:
#             return node.right
#         elif not node.right:
#             return node.left
#         else:
#             temp = node.right
#             while temp.left: temp = temp.left
#             temp.left = node.left
#             return node.right

#     def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
#         if not root:
#             return root
#         if root.val > key:
#             root.left = self.deleteNode(root.left, key)
#         elif root.val < key:
#             root.right = self.deleteNode(root.right, key)
#         else:
#             return self.deleteOneNode(root)
#         return root

class Solution:
    def deleteOneNode(self, node):
        if not node.left and not node.right:
            return None
        elif not node.left:
            return node.right
        elif not node.right:
            return node.left
        else:
            temp = node.right
            while temp.left: temp = temp.left
            temp.left = node.left
            return node.right

    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root: return root
        cur = root
        pre = None
        while cur:
            if cur.val == key:
                break
            pre = cur
            if cur.val > key:
                cur = cur.left
            else:
                cur = cur.right
        if not cur: return root
        if not pre: return self.deleteOneNode(root)
        if pre.left and pre.left.val == key: pre.left = self.deleteOneNode(cur)
        if pre.right and pre.right.val == key: pre.right = self.deleteOneNode(cur)
        return root

File: C++/450. Delete Node in a BST.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     TreeNode* deleteNode(TreeNode* root, int key) {
 //         if (!root) return root;
 //         if (root->val < key) root->right = deleteNode(root->right, key);
 //         else if (root->val > key) root->left = deleteNode(root->left, key);
 //         else{
 //             if (!root->left && !root->right) return NULL;
 //             else if (!root->left) return root->right;
 //             else if (!root->right) return root->left;
 //             else{
 //                 TreeNode* cur = root->right;
 //                 TreeNode* parent = NULL;
 //                 while (cur){
 //                     parent = cur;
 //                     cur = cur->left;
 //                 }
 //                 parent->left = root->left;
 //                 return root->right;
 //             }
 //         }
 //         return root;
 //     }
 // };

class Solution {
public:
    TreeNode* deleteOneNode(TreeNode* node) {
        if (!node->left && !node->right) return NULL;
        else if (!node->left) return node->right;
        else if (!node->right) return node->left;
        else {
            TreeNode* temp = node->right;
            while (temp->left) temp = temp->left;
            temp->left = node->left;
            return node->right;
        }
    }
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (!root) return root;
        TreeNode* cur = root;
        TreeNode* parent = NULL;
        while (cur) {
            if (cur->val == key) break;
            parent = cur;
            if (cur->val > key) cur = cur->left;
            else cur = cur->right;
        }
        if (!cur) return root;
        if (!parent) return deleteOneNode(cur);
        if (parent->left && parent->left->val == key) parent->left = deleteOneNode(cur);
        if (parent->right && parent->right->val == key) parent->right = deleteOneNode(cur);
        return root;
    }
};